heartbeat/monitors/util.go

```diff
@@ -379,66 +343,53 @@ func resolveIPEvent(host, ip string, rtt time.Duration) common.MapStr {
 	}
 }
 
-func resolveErr(host string, err error) (common.MapStr, []TaskRunner, error) {
-	event := common.MapStr{
-		"monitor": common.MapStr{
-			"host": host,
-		},
-		"resolve": common.MapStr{
-			"host": host,
+func resolveErr(host string, err error) (*beat.Event, []Job, error) {
+	event := &beat.Event{
+		Fields: common.MapStr{
+			"monitor": common.MapStr{
+				"host": host,
+			},
+			"resolve": common.MapStr{
+				"host": host,
+			},
 		},
 	}
+
 	return event, nil, err
 }
 
 // WithFields wraps a TaskRunner, updating all events returned with the set of
 // fields configured.
-func WithFields(fields common.MapStr, r TaskRunner) TaskRunner {
-	return MakeCont(func() (common.MapStr, []TaskRunner, error) {
-		event, cont, err := r.Run()
-		if event != nil {
-			event = event.Clone()
```

Comment:
I think this clone here is important to prevent races but I can't find it in the code anymore in MergeEventFields.

Reply:
Good catch.

Reply:
Let's make to double check this. Can you also check when this was introduced and the PR comments for it?

Reply:
Looks like it was introduced in https://github.com/elastic/beats/pull/6950 , which did this to prevent races among shared structures in the case continuations were created. After a good amount of spelunking I don't see any races.

However, reasoning about those races is really hard to do, you basically need to bootstrap all the code into your head and assume that you understand it correctly.

That said, I don't think the approach of fixing it in this spot is necessarily correct. This assumes that `WithFields` is the only place mutation happens.

I'd like to think about this for a bit and see if there's a more bullet-proof solution I can think of (short of moving toward immutable datastructures).

Reply:
There seem to be two concerns here:

1. Ensuring that each continuation works on a new instance of `beat.Event`.
2. Ensuring that data is not shared between those instances.

Solving 1 could be done by making `beat.Event` an argument, rather than a return type, to most runner functions. So `Run() (*beat.Event, []Job, error)` would turn into `Run(*beat.Event) ([]Job, error)`. This means we can instantiate a new Event when we spawn a new continuation.

Solving 2 is harder. It means ensuring that any updates to the event are not shared. `WithFields` already does a deep copy to try to achieve this (including in this PR). My sense is that a code audit to ensure all mutations are done with `WithFields` is sufficient. I did notice that the deep copy used there does not handle arrays. Luckily I do not believe we use arrays yet anywhere. 

One other solution to 2 would be to wrap `beat.Event` into a new type that only had accessors  that made threadsafe copies of passed in data, and had an optional `PutUnsafe` or similar to let us avoid doing a ton of unnecessary copies when possible.

Thoughts @urso and @ruflin ?

Reply:
What you see as disadvantages of 1?

Reply:
To be clear, I don't think 1 and 2 are competing proposals. These are separate issues with separate solutions. I don't see any disadvantages with that strategy aside from it always being a bit awkward for functions to mutate their arguments.

Reply:
What is your suggestion to move forward and not reintroduce the potential race?

Reply:
My suggestion is to re-introduce the clone in this PR and then introduce a subsequent PR that restructures things so as to make it unnecessary. That would basically mean rewriting all the function signatures for everything here. I got pretty far with it on a [new branch](https://github.com/andrewvc/beats/tree/unify-experiment) but I don't think it's worth the effort right at this moment, and I think it'd make this PR far too unwieldy to review.

Reply:
+1 on your approach. Like this we can also discuss the follow up changes separately.